 /**:
  ros__parameters:
  ############ system configuration ############
    log_to_file: true
    log_debug: true
    log_enabled:
      - navigation
      #- radar.pipeline
      # - navigation.sensor_input
      - radar.online_converter
      #- radar.pc_extractor
      #- radar.odometry_icp
      #- radar.odometry_gyro
      #- radar.odometry_preintegration
      # - radar.localization_icp
      # - tactic
      # - tactic.pipeline
      - tactic.module
      #- mpc.debug
      #- cbit.control
      #- test
      #- mission.state_machine
      #- mission.server
    robot_frame: base_link
    env_info_topic: env_info
    radar_frame: radar_link
    radar_topic: /radar/b_scan_msg
    gyro_frame: MTi30_link
    gyro_topic: /vn100/data_raw
    gyro_bias: -0.027524760208093454

    queue_size: 5
    graph_projection:
      origin_lat: 47.322101 # UTIAS: 43.78220
      origin_lng: -71.147739 # UTIAS: -79.4661
      origin_theta: 0.0
      scale: 1.0
    # Added Tactical Params
    tactic:
      enable_parallelization: true
      preprocessing_skippable: false
      odometry_mapping_skippable: false
      localization_skippable: false
      task_queue_num_threads: 1
      task_queue_size: -1

      route_completion_translation_threshold: 0.1

      chain:
        min_cusp_distance: 1.5
        angle_weight: 7.0
        search_depth: 5
        search_back_depth: 10
        distance_warning: 5.0
      save_odometry_result: true
      save_localization_result: true
      visualize: true
      rviz_loc_path_offset:
        - 0.0
        - 0.0
        - 0.0                                                                                                                                                                  
    ############ pipeline configuration ############
    pipeline:
      type: radar
      preprocessing: 
        - conversion
        - extraction
        - filtering
      odometry:
        #- gyro # you can turn off gyro and preintegration here
        - icp
        - mapping
        - vertex_test
        - memory
      localization: 
        - recall
        - icp
        - memory
      submap_translation_threshold: 1.5
      submap_rotation_threshold: 30.0
      save_raw_point_cloud: false
      save_radar_images: false
    preprocessing:
      conversion:
        type: radar.online_converter
        radar_resolution: 0.043809514
        encoder_bin_size: 5600 # encoder bin size
      extraction:
        type: radar.pc_extractor
        detector: kpeaks #modified_cacfar # choose detector \in (kstrongest, cen2018, cacfar, oscfar, modified_cacfar,caso_cfar)
        minr: 1.0 # mininum detection distance
        maxr: 100.0 # maximum detection distance
        # TODO: make cart_resolution an integer mult of radar_resolution
        cart_resolution: 0.2384 # m/pixel of BEV Cartesian radar image
        beta: 0.0 # Doppler correction factor (set to 0 to turn this off)
        range_offset: -0.319910 # RAS3 range offset in meters -0.220240  -0.3088
        kpeaks:
          kstrong: 4
          threshold2: 0.0
          threshold3: 0.25
        kstrongest:
          kstrong: 4 # k-strongest per azimuth to retain was 12
          static_threshold: 0.3
        cen2018:
          zq: 2.5
          sigma: 8
        cacfar:
          width: 41
          guard: 2
          threshold: 1.0
          threshold2: 0.0 # cfar: mult of mean, T = threshold * Z + threshold2 * mean + threshold3
          threshold3: 0.08
        oscfar:
          width: 41
          guard: 0
          kstat: 20 # used in oscfar, kth statistic to return in window
          threshold: 1.0
          threshold2: 0.0 # cfar: mult of mean, T = threshold * Z + threshold2 * mean + threshold3
          threshold3: 0.07
        modified_cacfar:
          width: 137 # Old: 101, New: 41 # Elliot 101
          guard: 7 # Old: 5, New: 2 # 5
          threshold: 0.48 # Old: 0.2, New: 1.0 # was 0.7
          threshold2: 0.0 # Old: 4.0, New: 0.0 cfar: mult of mean, T = threshold * Z + threshold2 * mean + threshold3
          threshold3: 0.23 # Old: 0.0, New: 0.08 # was 0.23
        caso_cfar:
          width: 101 # was 101
          guard: 5 # was 5
          threshold: 800.0 # I may have to change 100,200,400,600 ... 800
        visualize: true

      filtering:
        type: radar.preprocessing

        num_threads: 8

        voxel_downsample: false
        frame_voxel_size: 1.0

        window_size: 3.0 # number of neighboring azimuths used to compute line normal
        azimuth_res: 0.015707963267948967 # azimuth resolution in radians (0.1 in degree), giveng M=400 2pi/400 = 0.015707963267948967
        rho_scale: 1.0 # when finding neighbors, set rho to log(rho) / rho_scale

        num_sample_linearity: 5000 # max number of sample after filtering based on linearity
        min_linearity_score: -1.0 # min linearity score (negative value means no filtering)

        visualize: true

    odometry:       
      icp:
        type: radar.odometry_icp

        # continuous time estimation
        traj_num_extra_states: 0
        traj_qc_diag: [1.0, 0.001, 0.0001, 0.0001, 0.0001, 0.001]

        # ICP parameters # TOTUNE
        num_threads: 10 # 8
        first_num_steps: 2 # 2
        initial_max_iter: 4 # 4
        initial_max_pairing_dist: 1.5 # 5.0 # now same as lidar
        initial_max_planar_dist: 1.0 # 5.0
        refined_max_iter: 15 # 50
        refined_max_pairing_dist: 1.0 # 5.0
        refined_max_planar_dist: 0.3 # 5.0
        averaging_num_steps: 5 # 5
        huber_delta: 1.0 # 1.0
        cauchy_k: 0.2 # was 0.8
        remove_orientation: false

        # steam parameters #TOTUNE
        verbose: false
        max_iterations: 1
        # gyro_cov: 1e-4

        # threshold #TOTUNE
        min_matched_ratio: 0.40

        visualize: true

      mapping:
        type: radar.odometry_map_maintenance

        map_voxel_size: 0.5

        crop_range_front: 80.0
        back_over_front_ratio: 1.0

        point_life_time: 5.0

        visualize: true

      vertex_test:
        type: radar.vertex_test

        max_translation: 0.4
        max_rotation: 10.0

      memory:
        type: "live_mem_manager"

        window_size: 5 # unload data older than this many keyframes

    localization:
      recall:
        type: radar.localization_map_recall

        map_version: "pointmap"

        visualize: true

      icp:
        type: radar.localization_icp

        use_pose_prior: true
        # ICP parameters #TO TUNE
        num_threads: 10 # 8
        first_num_steps: 4 # 2
        initial_max_iter: 8 # 4
        initial_max_pairing_dist: 10.0 # 5.0
        initial_max_planar_dist: 1.0 # 5.0
        refined_max_iter: 10 # 50
        refined_max_pairing_dist: 1.0 # 5.0
        refined_max_planar_dist: 0.3 # 5.0
        averaging_num_steps: 5 # 5
        huber_delta: 1.0 # 1.0
        cauchy_k: 0.2

        # steam parameters #TOTUNE
        verbose: false
        max_iterations: 3

        # threshold #TOTUNE
        min_matched_ratio: 0.30

      memory:
        type: "graph_mem_manager"

        vertex_life_span: 5
        window_size: 3

    path_planning:
      type: stationary